api_app/serializers.py

```diff
@@ -226,13 +449,14 @@ def validate(self, attrs: dict) -> dict:
             attrs["observable_name"] = attrs["observable_name"].lower()
         # calculate ``md5``
         attrs["md5"] = calculate_md5(attrs["observable_name"].encode("utf-8"))
+        self.filter_analyzers_and_connectors(attrs)
```

Comment:
That's what I had initially planned. However, that can't be done as for filter_analyzers_and_connectors depends on attrs["file_mimetype"], which is not set until FileAnalysisSerializer.validate runs.

Reply:
So wait, we now have the method `filter_analyzers_and_connectors` in the class `_AbstractJobCreateSerializer` that checks "fields" (i.e. serialized_data["file_mimetype"]) that could not be set? You see that this logically makes no sense.
My approach is to actually remove that check in `filter_analyzers_and_connectors`, and in the class `FileAnalysisSerializer` override the method and _there_ add again this check

Reply:
If I override `filter_analyzers_and_connectors` in `FileAnalysisSerializer` and add the check there, won't I still have to keep the 2 calls to `self.filter_analyzers_and_connectors(attrs)` in child classes instead of parent class?

Reply:
No, you could just have it  called in the parent class and in the `FileAnalysisSerializer` call `super().validate()` **AFTER** the set up of `attrs["md5"]`

Reply:
No, you would need to leverage inheritance properly. After having removed the code related to the "file_mimetype" in the `filter_analyzers_and_connectors` you can add it in the overwrite of the same method in the `FileAnalysisSerializer` class:
```
def filter_connectors(...):
    super(FileAnalysisSerializer, ...
    # check mimetype here
```
In this way, first, you leverage the common code, then you perform the check related to files only. This approach could be applied also for observables (`serialized_data["observable_classification"]`)

Reply:
```python3
class _AbstractJobCreateSerializer(rfs.ModelSerializer):
    """
    Base Serializer for ``Job`` model's ``create()``.
    """

    tags_labels = rfs.ListField(default=list)
    runtime_configuration = rfs.JSONField(required=False, default={}, write_only=True)
    analyzers_requested = rfs.ListField(default=list)
    connectors_requested = rfs.ListField(default=list)
    md5 = rfs.HiddenField(default=None)

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.warnings = []

    def validate(self, attrs: dict) -> dict:
        # check and validate runtime_configuration
        runtime_conf = attrs.get("runtime_configuration", {})
        if runtime_conf and isinstance(runtime_conf, list):
            runtime_conf = json.loads(runtime_conf[0])
        attrs["runtime_configuration"] = runtime_conf
        -------------------- HERE ----------------------
        self.filter_analyzers_and_connectors(attrs)
        return attrs

    @staticmethod
    def filter_analyzers(serialized_data: Dict, warnings: List) -> List[str]:
        # the same but removed the check on file

class FileAnalysisSerializer(_AbstractJobCreateSerializer):
    @staticmethod
    def filter_analyzers(serialized_data: Dict, warnings: List) -> List[str]:
        # added the check on the file
        
    def validate(self, attrs: dict) -> dict:
        logger.debug(f"before attrs: {attrs}")
        # calculate ``file_mimetype``
        attrs["file_mimetype"] = calculate_mimetype(attrs["file"], attrs["file_name"])
        # calculate ``md5``
        file_obj = attrs["file"].file
        file_obj.seek(0)
        file_buffer = file_obj.read()
        attrs["md5"] = calculate_md5(file_buffer)
        ---------------------------------- HERE --------------------------
        attrs = super(FileAnalysisSerializer, self).validate(attrs)
        logger.debug(f"after attrs: {attrs}")
        return attrs

Reply:
In `filter_analyzers_and_connectors`, there is a code snippet of this type:
```
if this_is_a_file:
    # Perform some file type checks (including file_mimetype
else:
    # Perform some observable type checks
```
There are 2 approaches I can follow here.
1) Leave this code block as it is.
2) Remove both, override `filter_analyzers_and_connectors` in both child analyzers and add them there.

Both these approaches would work equally well. Which one would you prefer me to take?
@mlodic @0ssigeno

Reply:
Imho the secoond approach is the good one

Reply:
Done